### PR TITLE
Fix issues with applyMultiSignatures

### PR DESCRIPTION
**Pull request by rzabarazesh**
* Fixed signatures overwriting each other if applied one at a time
* Fixed the order of signatures not being checked against redeemScript. signatures will now be sorted regardless of their initial order.

### CODE REVIEW

File: src/transaction.js [hunk 1]
```diff
@@ -483,35 +500,67 @@ function applyMultiSignatures (
...
+      sigs = sigs.slice(2 + sig.length)
+    }
+    const newSigs = signatures.filter(sig => prevSigs.indexOf(sig) < 0)
+    allSigs = newSigs.concat(prevSigs)
+  }
 
-  // const signaturesFixed = pubKeys.map(pubKey => {
-  //   const keyPair = secp256k1.keyFromPublic(pubKey)
+  // Fix signature order
+  var unmatched = JSON.parse(JSON.stringify(allSigs))
```
Comment (nicolam1) [inline src/transaction.js]: how come we are using a mix of es6 and regular js standards use const or let.

File: test/transaction.js [hunk 2]
```diff
@@ -108,11 +108,24 @@ it('multiSign() for x-of-5 should be deterministic', function () {
   var sig1 = zencashjs.transaction.multiSign(txobj, 0, privKeys[0], redeemScript)
   var sig2 = zencashjs.transaction.multiSign(txobj, 0, privKeys[1], redeemScript)
   var tx0 = zencashjs.transaction.applyMultiSignatures(txobj, 0, [sig1, sig2], redeemScript)
+  var tx0Serialized = zencashjs.transaction.serializeTx(tx0)
 
-  // Serialize the transaction
-  var serializedTx = zencashjs.transaction.serializeTx(tx0)
+  var txObjSig1 = zencashjs.transaction.applyMultiSignatures(txobj, 0, [sig1], redeemScript)
```
Comment (nicolam1) [inline test/transaction.js]: const?

**Comment by nicolam1** [comment]
@cronicc do we have ci for this repo?

**Comment by cronicc** [comment]
Yes, we do

**Comment by nicolam1** [comment]
for some reason the check is not reported then.

**Comment by cronicc** [comment]
https://travis-ci.org/ZencashOfficial/zencashjs/builds/569927719?utm_source=github_status&utm_medium=notification

The repo still seems to be using the travis legacy open source integration. So it's running on travis.ci-org instead of travis-ci.com,

**Comment by al73rna** [comment]
The pipeline was failing because I hadn't committed the output code as well. It's fixed now.

Also addressed your comments @nicolam1

File: src/transaction.js [hunk 3]
```diff
@@ -460,6 +460,23 @@ function multiSign (
   return getScriptSignature(privKey, signingTx, hashcode)
 }
 
+function txToMsg(_txObj, i, redeemScript, hashcode) {
+  var txObj = JSON.parse(JSON.stringify(_txObj))
```
Comment (nicolam1) [inline src/transaction.js]: const

Comment (al73rna) [reply src/transaction.js]: done

**Comment by cronicc** [comment]
Not a regression but an additional bug, zencashjs doesn't handle m-of-n multisig correctly if m=n.

**Comment by Rifat0000** [comment]
What

**Comment by brutuscat** [comment]
Hi, I'm looking into multisig wallets for ZEN and this JS lib looks like a great programmatic way of using this feature. Are the issues reported above still relevant? Should use something else instead for ZEN MultiSig wallets?

**Comment by angievi** [comment]
> Hi, I'm looking into multisig wallets for ZEN and this JS lib looks like a great programmatic way of using this feature. Are the issues reported above still relevant? Should use something else instead for ZEN MultiSig wallets?

This PR would extent the multisignature functionality of zencashjs, but even without this PR multisignature functionality works fine.

please take a look at the examples in https://github.com/HorizenOfficial/zencashjs/blob/master/README.md#example-usage-multi-sig

**Review by QAWSED4455 — CHANGES_REQUESTED** [review]
issue27/applyMultiSignatures-fix

**Pull request closed without merging**